include/aos/common/tools/algorithm.hpp

```diff
@@ -0,0 +1,295 @@
+    virtual I  end(void)         = 0;
+    virtual CI begin(void) const = 0;
+    virtual CI end(void) const   = 0;
+
+    /**
+     * Removes item from container.
+     *
+     * @param item item to remove.
+     * @return RetWithError<I> pointer to next after deleted item.
+     */
+    virtual RetWithError<I> Remove(I item) = 0;
+
+    /**
+     * Checks if container is empty.
+     *
+     * @return bool.
+     */
+    bool IsEmpty() const { return Size() == 0; }
+
+    /**
+     * Checks if container is full.
+     *
+     * @return bool.
+     */
+
+    bool IsFull() const { return Size() == MaxSize(); }
+
+    /**
+     * Returns true if container is not empty.
+     *
+     * @return bool.
+     */
+    operator bool() const { return Size() > 0; }
+
+    /**
+     * Checks if container equals to another container.
+     *
+     * @param container to compare with.
+     * @return bool.
+     */
+    template <typename C>
+    bool operator==(const C& container) const
+    {
+        if (container.Size() != Size()) {
+            return false;
+        }
+
+        for (auto it1 = begin(), it2 = container.begin(); it1 != end(); it1++, it2++) {
+            if (!(*it1 == *it2)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Checks if container doesn't equal to another container.
+     *
+     * @param container to compare with.
+     * @return bool.
+     */
+    template <typename C>
+    bool operator!=(const C& container) const
+    {
+        return !operator==(container);
+    }
+
+    /**
+     * Returns container first item.
+     *
+     * @return RetWithError<T&>.
+     */
+    RetWithError<T&> Front()
+    {
+        if (IsEmpty()) {
+            return {*end(), ErrorEnum::eNotFound};
+        }
+
+        return *begin();
+    }
+
+    /**
+     * Returns container first const item.
+     *
+     * @return RetWithError<const T&>.
+     */
+    RetWithError<const T&> Front() const
+    {
+        if (IsEmpty()) {
+            return {*end(), ErrorEnum::eNotFound};
+        }
+
+        return *begin();
+    }
+
+    /**
+     * Returns container last item.
+     *
+     * @return RetWithError<T&>.
+     */
+    RetWithError<T&> Back()
+    {
+        if (IsEmpty()) {
+            return {*end(), ErrorEnum::eNotFound};
+        }
+
+        auto it = end();
+
+        return *(--it);
+    }
+
+    /**
+     * Returns container last const item.
+     *
+     * @return RetWithError<const T&>.
+     */
+    RetWithError<const T&> Back() const
+    {
+        if (IsEmpty()) {
+            return {*end(), ErrorEnum::eNotFound};
+        }
+
+        auto it = end();
+
+        return *(--it);
+    }
+
+    /**
+     * Finds const element in container.
+     *
+     * @param item element to find.
+     * @return RetWithError<CI>.
+     */
+    RetWithError<CI> Find(const T& item) const
+    {
+        for (auto it = begin(); it != end(); it++) {
```

Comment:
Consider using pre-increment for all these loop iterators.

Reply:
done